### ISSUE TITLE
add an optional `crate_name` attr to `rust_library`

### ISSUE THREAD

**Issue by mfarrugi** [bazelbuild/rules_rust#220]
Imagine we have a protocol defined somewhere, and we want generate polyglot client libraries, eg
`/some/protocol/BUILD`
```
proto_library(
    name = "protocol",
    ...
)

java_proto_library(
    name = "java",
    ...
)
rust_proto_library(
    name = "rust", 
    ...
)
```

This doesn't work well for rust libraries since the crate ends up being named `rust`, and it's easy to create conflicts.

I think we should add an optional `crate_name` attr to `rust_library`.

As a kind of tangent, it might be useful to revisit the label -> crate name mapping..

**Comment by acmcarther** [comment]
The issue with lib.name vs crate_name is definitely a real issue. I agree there are enough cases to merit this feature. I think this was raised elsewhere and I remember an extended discussion on the issue, but I don't have a link readily available.

The suggestion "crate_name" SGTM.

---

That said, I do have a suggestion for your particular problem. Generally, @ goog, we name our proto libraries after the original lib. An example:

```python
proto_library(
    name = "my_proto",
    srcs = ["my.proto"],
    deps = [
    ],
)

cc_proto_library(
    name = "my_cc_proto",
    deps = [":my_proto"],
)

go_proto_library(
    name = "my_go_proto",
    deps = [":my_proto"],
)

java_proto_library(
    name = "my_java_proto",
    deps = [":my_proto"],
)
```

I think this can be extended to the rust rule (though I'm unfamiliar with it) as
```python
rust_proto_library(
    name = "my_rust_proto",
    deps = [":my_proto"],
)
```

Perhaps you have a compelling reason to avoid the name prefix, though.

**Comment by mfarrugi** [comment]
I am not attached to any naming scheme. At a glance, I know at least Java
doesn't couple the name of the jar with the import path.. not as familiar
with the rest.

I'll look to add the parameter.

**Comment by dayfine** [comment]
somewhat related to this problem:

Suppose I have two proto files that defines `Foo` and `Bar` respectively, and `Bar` depends on `Foo`.
```python
# package my.mod;
# message Foo { ... }
proto_library(
    name = "foo_proto",
    srcs = ["foo.proto"],
    deps = ["@com_google_protobuf//:timestamp_proto"],
)

rust_proto_library(
    name = "foo_rust_proto",
    deps = [":foo_proto"],
)

# package my.mod;
# message Bar {
#     Foo foo = 1;
# }
proto_library(
    name = "bar_proto",
    srcs = ["bar.proto"],
    deps = [":foo_proto"],
)

rust_proto_library(
    name = "bar_rust_proto",
    deps = [":bar_proto"],
)
```

In C++ (my daily language), the generated code I would use are `my::mod::Foo` and `my::mod::Bar`, which is easy to understand.

However with rust, I seem to get something like:
```rust
extern crate bar_rust_proto;

use self::bar_rust_proto::*;

// I can use both `Foo` and `Bar` here.
```
Is this the `crate_name` problem we are talking about? So my impression is both `foo_rust_proto` target and `bar_rust_proto` will generate a separate *crate* and contain their own declaration(?) of `pub mod foo`, which is confusing if not problematic. 

I suppose the `crate_name` would need to behave similarly to the `java_package`, right?

*Actually I do have this question:* is it correctly in my example to have two separate `rust_proto_library` that would include the same types in different crates? Or should I use one proto target (which I doubt work well when there are many protos with complex dependencies)?

**Comment by UebelAndre** [comment]
I believe this was implemented in https://github.com/bazelbuild/rules_rust/pull/645. @illicitonion would you concur?

**Comment by illicitonion** [comment]
Looks done to me!
